conan/tools/cmake/toolchain/blocks.py

```diff
@@ -667,8 +667,63 @@ class TryCompileBlock(Block):
         """)
 
 
-class GenericSystemBlock(Block):
+class CompilersBlock(Block):
     template = textwrap.dedent(r"""
+        {% if compiler %}
+        if(NOT DEFINED ENV{CC})
+        set(CMAKE_C_COMPILER {{ compiler }})
+        endif()
+        {% endif %}
+        {% if compiler_cpp %}
+        if(NOT DEFINED ENV{CXX})
```

Comment:
I would like the opinion of @czoido and @jcar87 about this (I know it was already there ,not this PR).
If Conan is controlling the CMAKE_CXX_COMPILER, maybe that should still be prioritized against a general environment CXX env-var? Not sure about this.

Reply:
But if I'm not wrong the current priority as this PR is would be:
- `tools.build:compilers `configuration
- environment variables
- the result from `_get_windows_compilers`
So it should be already prioritized.
I was also wondering if now that we have the conf we still need to set to set these values using the  `_get_windows_compilers` functions?

Reply:
I am not sure. If ``ENV{CXX}`` is defined, it will not generate the ``CMAKE_CXX_COMPILER`` entry in the toolchain at all.

> I was also wondering if now that we have the conf we still need to set to set these values using the _get_windows_compilers functions?

I think it is nice to have those, because those are the default needed to work with those setups, and it can be extra burden on users to specify the ``cl.exe`` compiler when they are targeting the ``compiler=msvc`` setting, the VS in installed correctly, just because the setup is via ``Ninja`` instead of other CMake generator.

Reply:
> I am not sure. If `ENV{CXX}` is defined, it will not generate the `CMAKE_CXX_COMPILER` entry in the toolchain at all.
> 

But the `CMAKE_LANG_COMPILER` from the conf is set below in this code:

```
        {% for lang, compiler_path in compilers_by_conf.items() %}
        set(CMAKE_{{ lang|upper }}_COMPILER "{{ compiler_path|replace('\\', '/') }}")
        {% endfor %}
```

I think probably the logic should be changed, because the conf will override the other values even if you are not setting them in the conf.

> > I was also wondering if now that we have the conf we still need to set to set these values using the _get_windows_compilers functions?
> 
> I think it is nice to have those, because those are the default needed to work with those setups, and it can be extra burden on users to specify the `cl.exe` compiler when they are targeting the `compiler=msvc` setting, the VS in installed correctly, just because the setup is via `Ninja` instead of other CMake generator.

👍

Reply:
Sorry, I've added an extra if-else statement to keep a clear structure in the jinja template